external-crates/move/crates/move-vm-runtime/src/loader/arena.rs

```diff
@@ -0,0 +1,76 @@
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+// Arena Definitions
+
+use std::mem::MaybeUninit;
+
+use bumpalo::Bump;
+
+pub struct Arena(Bump);
+
+impl Arena {
+    pub fn new() -> Self {
+        Arena(Bump::new())
+    }
+
+    /// SAFETY: it is the caller's responsibility to ensure that `self` is not shared across
+    /// threads during this call. This should be fine as the loader should run in lock-isoltion and
+    /// nothing should allocate into a LoadedModule after it is loaded.
+    pub fn alloc_slice<T>(&self, items: impl ExactSizeIterator<Item = T>) -> *mut [T] {
+        let slice = self.0.alloc_slice_fill_iter(items);
+        slice as *mut [T]
+    }
+}
+
+// SAFETY: these are okay, if callers follow the documented safety requirements for `Arena`'s
+// unsafe methods.
+unsafe impl Send for Arena {}
+unsafe impl Sync for Arena {}
+
+/// Returns a pointer to a slice, but nulled. This must be set before use.
+pub fn null_ptr<T>() -> *const [T] {
+    unsafe { MaybeUninit::<*const [T]>::zeroed().assume_init() }
+}
+
+pub fn ref_slice<'a, T>(value: *const [T]) -> &'a [T] {
+    unsafe { &*value }
+}
+
+pub fn mut_to_ref_slice<'a, T>(value: *mut [T]) -> &'a [T] {
+    unsafe { &*value }
+}
+
+pub fn to_mut_ref_slice<'a, T>(value: *mut [T]) -> &'a mut [T] {
+    unsafe { &mut *value }
+}
+
+pub fn to_ref<'a, T>(value: *const T) -> &'a T {
+    unsafe { &*value as &T }
+}
+
+#[derive(Clone, Copy)]
+pub struct ArenaPointer<T>(*const T);
+
+impl<T> ArenaPointer<T> {
```

Comment:
I am thinking this is what should "travel" across calls and everywhere.
We could then use `to_ref` when it's convinient to pass a reference and `to_const` in very few places where "access to code" is required.
I think that is mostly what we have but I am asking if that is how you see things

Reply:
Todd's suggestion was to do use `*const Function` everywhere initially and introduce a wrapper as a follow-up. I also prefer the `ArenaPointer<T>`, so I will make this change.